### DISCUSSION TITLE
Recursive conditional updates

### DISCUSSION THREAD

**Discussion by kpdemetriou** [geldata/gel#1952] [Q&A] [answered]
Hi EdgeDB team, awesome project you got going on here. I was wondering if one could do something akin to the following. This fails with `INSERT statements cannot be used inside conditional expressions` and `cannot drop object type 'playground::Address' because other objects in the schema depend on it` but perhaps there's an alternative way to express it...?

```
WITH MODULE playground
UPDATE Person
FILTER Person.name = <str>$name
SET {
    address := (
        (
            INSERT Address {
                street := <str>$street,
                country := <str>$country
            }
        ) IF NOT EXISTS Person.address ELSE (
            UPDATE Address
            FILTER Address = Person.address
            SET {
                street := <str>$street,
                country := <str>$country
            }
        )
    )
}
```

**Comment by elprans** [comment, marked as answer]
Hi @kpdemetriou!  This is an interesting case.  EdgeDB currently supports the `INSERT .. UNLESS CONFLICT` syntax that works similarly to PostgreSQL's `ON CONFLICT`, however there is currently no way to condition the conflict clause on the properties of a _relationship_.  We have some considerations about adding support for that, but for now the best way to handle cases like this is to reverse the relationship and use the existing `UNLESS CONFLICT` feature: 

```
CREATE TYPE Person {
    CREATE PROPERTY name -> str;
};

CREATE TYPE Address {
    CREATE REQUIRED LINK person -> Person {
        # This is required for UNLESS CONFLICT
        # and makes the user-address relationship
        # proper.
        CREATE CONSTRAINT exclusive;
    };
    
    CREATE PROPERTY street -> str;
    CREATE PROPERTY country -> str;
};

INSERT Person {
   name := 'John',
};

INSERT Address {
   person := (SELECT Person FILTER .name = 'John'),
   street := 'Dawning'
} 
UNLESS CONFLICT ON (.person)
ELSE (
   UPDATE Address SET { street := 'Dawning' }
);
```

For convenience, you can add a computed link to the `Person` type to access the address more naturally:

```
ALTER TYPE Person {
    CREATE LINK address := .<person[IS Address];
};
```

**Reply by kpdemetriou** [reply]
Hi @elprans, thanks for taking the time to respond! This solves the issue but also makes insertion take 2 statements. The motivation behind the question was to understand to what extent EdgeDB can behave like a document database in a relatively simple nested object scenario. Am I correct in understanding that this isn't fully supported at present?

**Reply by elprans** [reply]
That's right, although it's not a fundamental limitation, just a missing bit of implementation.  I opened #1960 to track.
